### PR TITLE
Re-write in Swift. 

### PR DESCRIPTION
**Pull request by benski**

### CODE REVIEW

File: Classes/Promise.swift [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+                wrappedBlock()
+            } else {
+                self.onCancel = wrappedBlock
+            }
+        }
+    }
+    
+    func cancel() {
+        cancelled = true
+        OSMemoryBarrier()
```
Comment (mvyrko) [inline Classes/Promise.swift:75]: Why do we need it here? 
as far as I see OSMemoryBarrier is deprecated. 
`@available(iOS, deprecated: 10.0, message: "Use atomic_thread_fence() from <stdatomic.h> instead")`

File: Classes/Promise.swift [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+            self.cancellationToken = cancellationToken
+        }
+        var shouldKeepPromise: Bool {
+            get {
+                return done != nil || always != nil || rejected != nil
+            }
+        }
+        
+        private func internalCall(with object: PromiseResult<ValueType>) {
+            if !self.cancellationToken.cancelled {
```
Comment (mvyrko) [inline Classes/Promise.swift]: can call it without self just if `!cancellationToken.cancelled`

File: Classes/Promise.swift [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+                internalCall(with: object)
+            }
+        }
+    }
+    
+    @discardableResult func then(_ onFulfilled: ((ValueType) -> Void)? = nil,
+                                 observed: ((ValueType) -> Void)? = nil,
+                                 rejected : ((Error) -> Void)? = nil,
+                                 always : (() -> Void)? = nil,
+                                 queue : DispatchQueue) -> PromiseCancelToken {
```
Comment (mvyrko) [inline Classes/Promise.swift]: Maybe make a sense use main queue as default parameter

File: Classes/Promise.swift [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+
+public class Promise<ValueType> : PromiseCancelToken {
+    
+    var fulfilledObject: PromiseResult<ValueType>?
+    
+    public class PromiseBlock {
+        var done : ((ValueType) -> Void)?
+        var observed : ((ValueType) -> Void)?
+        var rejected : ((Error) -> Void)?
+        var always : (() -> Void)?
```
Comment (mvyrko) [inline Classes/Promise.swift]: Looks like would be comfortable using typealiases for block types like PromiseErrorBlock, Empty and etc

File: Classes/Promise.swift [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+                    blocks.done = nil
+                    blocks.observed = nil
+                    blocks.rejected = nil
+                    blocks.always = nil
+                    
+                    var strongCount = false
+                    for block in `self`.blocks {
+                        if block.shouldKeepPromise {
+                            strongCount = true
+                            break
```
Comment (mvyrko) [inline Classes/Promise.swift]: from my point of view would be more readable one line conditional
`for block in blocks where block.shouldKeepPromise`

Comment (mvyrko) [reply Classes/Promise.swift]: More swifty would be 
```
if blocks.contains(where: { $0.shouldKeepPromise }) {
     `self`.cancel()
}
```

File: Classes/Promise.swift [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+            }
+        }
+    }
+
+}
+
+// fulfillment
+extension Promise {
+    
+    func fulfill(with: PromiseResult<ValueType>) {
```
Comment (mvyrko) [inline Classes/Promise.swift]: about name of func, in my view, 
your variant of func  name above`call(with object: PromiseResult<ValueType>) {` is better. 
here we can use something like:
` func fulfill(with result: PromiseResult<ValueType>)`

File: BAPromiseTests/PromiseBlocksTests.swift [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+        let alwaysExpect = XCTestExpectation()
+        promiseBlock.always = {
+            XCTAssertEqual(DispatchQueue.getSpecific(key: key), "testCallNonErrorOnQueue")
+            alwaysExpect.fulfill()
+        }
+        
+        promiseBlock.call(with: .failure(callObj))
+        
+        self.wait(for: [rejectedExpectation, alwaysExpect], timeout: 10)
+    }
```
Comment (mvyrko) [inline BAPromiseTests/PromiseBlocksTests.swift]: Looks like we can add more tests for validating callback queue when we use a chain of promises.
Case like:
- 1 queue: create promise and callbacks
- dispatch on 2 queue and transfer the promise to it 
- and new callbacks to the promise  
- fullfill promise on 3 queue
- validate

File: Classes/Promise.swift (hunk 1 above)
Comment (benski) [reply Classes/Promise.swift:75]: There was a specific scenario where we needed the memory barrier, but I can't remember it specifically. I'll look through and add a comment about it.
cancelled is read and written from multiple threads, but there's no "atomic" in Swift.

**Comment by mvyrko** [comment]
Can we implement tests that imitate livelocks/deadlocks? I have no good scenarios, for now, will provide as soon as I can.

File: Classes/Promise.swift [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+    
+    init() {
+        self = .success(())
+    }
+}
+
+public class PromiseCancelToken {
+    var cancelled: Bool = false
+    static let queue = DispatchQueue(label: "com.github.benski.promise")
+    var onCancel : (() -> Void)?
```
Comment (mvyrko) [inline Classes/Promise.swift]: Maybe we can use `public typealias Canceled = () -> Void`

File: Classes/Promise.swift [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+        return .success(())
+    }
+    
+    init() {
+        self = .success(())
+    }
+}
+
+public class PromiseCancelToken {
+    var cancelled: Bool = false
```
Comment (mvyrko) [inline Classes/Promise.swift]: Looks like it should be `private set `

File: Classes/Promise.swift [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+                } else {
+                    self.blocks.append(blocks)
+                }
+            }
+        }
+        
+        return cancellationToken
+    }
+    
+    lazy var blocks: Array<PromiseBlock> = []
```
Comment (mvyrko) [inline Classes/Promise.swift]: Looks like it should be `private` or `fileprivate`

File: Classes/Promise.swift [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+        cancellationToken = self.then({ value -> Void in
+            do {
+                let chained = try onFulfilled(value)
+                returnedPromise.fulfill(with: .success(chained))
+            } catch let error {
+                returnedPromise.fulfill(with: .failure(error))
+            }
+        }, queue: queue)
+        
+        return returnedPromise
```
Comment (mvyrko) [inline Classes/Promise.swift]: This function looks too similar with `func then<ReturnType>...`above
Can we remove `func map<ReturnType>`? if no then we can call `func then<ReturnType>...` with empty rejected block (ovveride delault)

File: Classes/Promise.swift [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+    }
+}
+
+// Join
+extension Promise {
+    // we can't currently have a templated extension (e.g. extension Array where Element = Promise<_>) so we have to do it this way
+    class func when(promises: [Promise<ValueType>]) -> Promise<Array<ValueType>> {
```
Comment (mvyrko) [inline Classes/Promise.swift]:
```
extension Array {

    func when<ValueType>() -> Promise<Array<ValueType>> where Element == Promise<ValueType> 
```

File: Classes/Promise.swift [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+                    token.cancel()
+                }
+            }, queue: Promise.queue)
+            cancelTokens.append(token)
+        }
+        
+        return returnedPromise
+    }
+    
+    class func join(promises: [Promise<ValueType>]) -> Promise<Array<PromiseResult<ValueType>>> {
```
Comment (mvyrko) [inline Classes/Promise.swift]:
```
extension Array {

    func join <ValueType>() -> Promise<Array<PromiseResult<ValueType>>> where Element == Promise<ValueType> 
```

File: Classes/Promise.swift [hunk 14]
```diff
@@ -264,24 +264,23 @@ extension Promise {
 }
 
 // Join
-extension Promise {
-    // we can't currently have a templated extension (e.g. extension Array where Element = Promise<_>) so we have to do it this way
-    class func when(promises: [Promise<ValueType>]) -> Promise<Array<ValueType>> {
-        guard promises.count > 0 else { return Promise<Array<ValueType>>([]) }
+extension Array {
+    func when<ValueType>() -> Promise<Array<ValueType>> where Element == Promise<ValueType> {
```
Comment (mvyrko) [inline Classes/Promise.swift]: 👍

**Review by mvyrko — APPROVED** [review]

**Comment by benski** [comment]
feedback addressed

**Pull request merged**